File: KmpDemo/iosApp/iosApp/CharacterView.swift
import SwiftUI
import Common

struct CharacterView: View {
    let state: SwApiUiState
    let character: Character?
    let onAppear: () -> Void
    
    var body: some View {
        VStack {
            switch state {
            case .fetchingNewCharacter(let characterId):
                if character?.id != nil && character!.id != characterId {
                    CharacterCard(character: character!)
                } else {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: .blue))
                        .scaleEffect(2)
                }
            case .characterLoaded:
                if character == nil {
                    Text("Something went wrong loading this character. Please try again later.")
                        .padding(.horizontal, 64)
                        .multilineTextAlignment(.center)
                } else {
                    CharacterCard(character: character!)
                }
            default:
                Text("Something went wrong loading the app. Please try again later.")
                    .padding(.horizontal, 64)
                    .multilineTextAlignment(.center)
            }
        }
        .onAppear { onAppear() }
    }
}

struct CharacterCard: View {
    let character: Character
    
    func cacheAndRender(image: Image, url: URL) -> some View {
        ImageCache.cacheImage(url: url)
        return image
            .resizable()
            .scaledToFill()
    }
    
    var body: some View {
        if #available(iOS 15.0, *) {
            let url = URL(string: "https://res.cloudinary.com/kalalau/kmpdemo/\(character.id).jpg")!
            
            VStack {
                if let cachedImage = ImageCache[url] {
                    Image(uiImage: cachedImage)
                        .resizable()
                        .scaledToFill()
                        .frame(width: 300, height: 300, alignment: .center)
                        .cornerRadius(10)
                } else {
                    AsyncImage(url: url, transaction: Transaction(animation: .easeInOut)) { phase in
                        switch phase {
                        case .success(let image):
                            cacheAndRender(image: image, url: url)
                        case .failure(_):
                            cacheAndRender(image: Image("empire-emblem"), url: url)
                        default:
                            Color.gray
                        }
                    }
                    .frame(width: 300, height: 300, alignment: .center)
                    .cornerRadius(10)
                }
                
                Spacer()
                    .frame(height: 10)
                
                Text(character.name)
                
                Spacer()
                    .frame(height: 10)
                
                Text("Appears in: \(character.appearsIn.joined(separator: ", "))")
                    .padding(.horizontal, 32)
            }
        } else {
            VStack {
                Image("empire-emblem")
                    .resizable()
                    .scaledToFill()
                
                Spacer()
                    .frame(height: 10)
                
                Text("Please upgrade to iOS 15")
            }
        }
    }
}

fileprivate class ImageCache {
    static private let cache = NSCache<NSString, UIImage>()
    
    static func cacheImage(url: URL) {
        let dataTask = URLSession.shared.dataTask(with: url) { data, _, error in
            guard let data = data, error == nil else {
                return
            }
            ImageCache.cache.setObject(UIImage(data: data) ?? UIImage(named: "empire-emblem")!, forKey: url.absoluteString as NSString)
        }
        dataTask.resume()
    }
    
    static subscript(url: URL) -> UIImage? {
        get {
            ImageCache.cache.object(forKey: url.absoluteString as NSString)
        }
    }
}

File: KmpDemo/androidApp/src/main/java/com/kalalaucantrell/kmpdemo/android/CharacterView.kt
package com.kalalaucantrell.kmpdemo.android

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import coil.request.ImageRequest
import com.kalalaucantrell.kmpdemo.android.SwApiUiState.CharacterLoaded
import com.kalalaucantrell.kmpdemo.android.SwApiUiState.FetchingNewCharacter
import com.kalalaucantrell.kmpdemo.network.Character

@Composable
fun CharacterView(state: SwApiUiState, character: Character?, onAppear: () -> Unit) {
    LaunchedEffect(onAppear) {
        onAppear()
    }

    Column {
        when (state) {
            is FetchingNewCharacter ->
                if (character?.id != null && state.characterId != character.id) {
                    CharacterCard(character)
                } else {
                    CircularProgressIndicator()
                }
            is CharacterLoaded ->
                if (character == null) {
                    Text(
                        "Something went wrong loading this character. Please try again later.",
                        modifier = Modifier
                            .padding(horizontal = 64.dp),
                        textAlign = TextAlign.Center
                    )
                } else {
                    CharacterCard(character)
                }
            else -> Text(
                "Something went wrong loading the app. Please try again later.",
                modifier = Modifier
                    .padding(horizontal = 64.dp),
                textAlign = TextAlign.Center
            )
        }
    }
}

@Composable
fun CharacterCard(character: Character) {
    Column {
        AsyncImage(
            model = ImageRequest.Builder(LocalContext.current)
                .data("https://res.cloudinary.com/kalalau/kmpdemo/${character.id}.jpg")
                .crossfade(300)
                .build(),
            placeholder = painterResource(R.drawable.ic_round_square),
            error = painterResource(R.drawable.ic_empire_emblem),
            contentDescription = "Image for character ${character.id}",
            contentScale = ContentScale.Crop,
            modifier = Modifier
                .align(Alignment.CenterHorizontally)
                .height(300.dp)
                .width(300.dp)
                .clip(RoundedCornerShape(10.dp))
        )
        Spacer(Modifier.height(10.dp))
        Text(
            character.name,
            modifier = Modifier.fillMaxWidth(),
            textAlign = TextAlign.Center
        )
        Spacer(Modifier.height(10.dp))
        Text(
            "Appears in: ${character.appearsIn.joinToString(", ")}",
            modifier = Modifier
                .padding(horizontal = 32.dp),
        )
    }
}
